### PR TITLE
ENH: Let initialisation from dicts use insertion order for python >= 3.6 (part III)

### LINKED ISSUE #19018: ENH: Let Series/DataFrame initialisation from dicts use insertion order when python>=3.6
Since Python 3.6 dicts have for practical purposes been ordered by insertion order in CPython, and from Cpython 3.7 dicts will be formally ordered by insertion order in the language specs.

Pandas currently orders series/dataframe comstructed from dict alphabetically. I propose to make the sort order dependent on python version, so Python <3.6 will keep the current behavior, while Python >= 3.6 will use insertion order.

I have played a bit with this and the code changes seem to be minimal (change a line in ``_init_dict`` in each of ``series.py``/``frame.py``). After making the changes only 36 tests are failing, which is good. Updating the tests will mainly consists of making the columns order fixed in places (i.e. add a ``columns=[...]`` for dataframes and ``index=[...]`` for series), so I predict this would not be a huge PR.

I could take this on, if there's agreement.

EDIT: Ok, ``sparse/frame.py`` will also have its ``_init_dict`` changed, but still I think this will be a small code change.

### PR DESCRIPTION
**Pull request by topper-123**
- [x] closes #19018
- [x] xref #19830 and #19859
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

This PR lets pandas allow initialisation using insertion order for python>=3.6. E.g. ``Series({'B': 1, 'A': 2})`` will retain order and not convert to ``Series({'A': 2, 'B': 1})``. As such this changes the API soemwhat fundamentally (but in a very good way, IMO).

Some questions:
* Should this change be at the top of the whatsnew, in order to increase visibility to this change?
* Should an option be added to ``pd.options``, so user can revert back to the old behaviour? I'm thinking here of users who have made tests based on dict sort-by-values, and whose tests will break by this new change
* Should there be a deprecation procedure, as asked by @gfyoung ?

I also need to look thorugh the documentation if changes should be made there, but would appreciate getting the basic PR approved first.

### CODE REVIEW

File: doc/source/whatsnew/v0.23.0.txt [hunk 1]
```diff
@@ -320,6 +320,57 @@ If installed, we now require:
...
+in these newer versions are ordered by insertion order
+(see also `PEP 468 <https://www.python.org/dev/peps/pep-0468/>`_).
+Pandas will from version 0.23 use insertion order, when creating series or
+data frames from dicts (:issue:`19018`) .
+
+Previous behaviour (and current behaviour if on Python < 3.6):
+
+.. code-block:: ipython
+
+  In [1]: pd.Series({'Income': 2000,
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: you need another space here

File: pandas/tests/frame/test_constructors.py [hunk 2]
```diff
@@ -241,6 +241,17 @@ def test_constructor_dict(self):
         # Corner cases
         assert len(DataFrame({})) == 0
```
Comment (jreback) [inline pandas/tests/frame/test_constructors.py]: make new tests

File: pandas/tests/series/test_constructors.py [hunk 3]
```diff
@@ -775,6 +775,17 @@ def test_constructor_dict(self):
         expected = Series([1, 2, nan, 0], index=['b', 'c', 'd', 'a'])
         assert_series_equal(result, expected)
```
Comment (jreback) [inline pandas/tests/series/test_constructors.py]: new tests

File: pandas/tests/series/test_constructors.py [hunk 4]
```diff
@@ -775,6 +775,17 @@ def test_constructor_dict(self):
...
 
+        # GH19018
+        # initialization ordering: by insertion order if python>= 3.6, else
+        # by value
+        d = {'b': 1, 'a': 0, 'c': 2}
```
Comment (jreback) [inline pandas/tests/series/test_constructors.py]: can you add ones for SparseSeries/DataFrame

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_constructors.py (hunk 2 above)
Comment (topper-123) [reply pandas/tests/frame/test_constructors.py]: Can you be more specific?

Comment (jreback) [reply pandas/tests/frame/test_constructors.py]: make these as new tests, don't add onto the existing ones (IOW give it a separate test name)

**Comment by topper-123** [comment]
Comments addressed.

File: doc/source/whatsnew/v0.23.0.txt [hunk 5]
```diff
@@ -240,7 +240,7 @@ The :func:`DataFrame.assign` now accepts dependent keyword arguments for python
   using ``.assign()`` to update an existing column. Previously, callables
   referring to other variables being updated would get the "old" values
 
-  Previous Behaviour:
+  Previous behaviour:
```
Comment (jschendel) [inline doc/source/whatsnew/v0.23.0.txt]: It looks like we generally default to American English spellings within the codebase, so could do "behavior" here instead.  Found many instances of "behaviour" in the codebase though, but still far outnumbered by "behavior".

File: doc/source/whatsnew/v0.23.0.txt [hunk 6]
```diff
@@ -253,7 +253,7 @@ The :func:`DataFrame.assign` now accepts dependent keyword arguments for python
       1  3 -2
       2  4 -3
 
-  New Behaviour:
+  New behaviour:
```
Comment (jschendel) [inline doc/source/whatsnew/v0.23.0.txt]: same

File: doc/source/whatsnew/v0.23.0.txt [hunk 7]
```diff
@@ -320,6 +320,57 @@ If installed, we now require:
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Until Python 3.6, dicts in Python had no formally defined ordering. Python
+version 3.6 and later have changed the ordering definition of dicts, so dicts
+in these newer versions are ordered by insertion order
+(see also `PEP 468 <https://www.python.org/dev/peps/pep-0468/>`_).
+Pandas will from version 0.23 use insertion order, when creating series or
+data frames from dicts (:issue:`19018`) .
+
+Previous behaviour (and current behaviour if on Python < 3.6):
```
Comment (jschendel) [inline doc/source/whatsnew/v0.23.0.txt]: same

File: doc/source/whatsnew/v0.23.0.txt [hunk 8]
```diff
@@ -320,6 +320,57 @@ If installed, we now require:
...
+   ...                 'Net result': 300})
+   Expenses     -1500
+   Income        2000
+   Net result     300
+   Taxes         -200
+   dtype: int64
+
+Note the series above is ordered alphabetically by the index values.
+
+New behaviour (for Python >= 3.6):
```
Comment (jschendel) [inline doc/source/whatsnew/v0.23.0.txt]: same

File: doc/source/whatsnew/v0.23.0.txt [hunk 9]
```diff
@@ -320,6 +320,57 @@ If installed, we now require:
...
+
+.. ipython:: python
+
+    pd.Series({'Income': 2000,
+               'Expenses': -1500,
+               'Taxes': -200,
+               'Net result': 300})
+
+Notice that the series is now ordered by insertion order. This new behaviour is
```
Comment (jschendel) [inline doc/source/whatsnew/v0.23.0.txt]: same

File: doc/source/whatsnew/v0.23.0.txt [hunk 10]
```diff
@@ -320,6 +320,57 @@ If installed, we now require:
...
+used for all relevant pandas types (``Series``, ``DataFrame``, ``SparseSeries``
+and ``SparseDataFrame``).
+
+If you wish to retain the old behaviour while using Python >= 3.6, you can use
```
Comment (jschendel) [inline doc/source/whatsnew/v0.23.0.txt]: same

File: doc/source/whatsnew/v0.23.0.txt [hunk 11]
```diff
@@ -320,6 +320,57 @@ If installed, we now require:
 | openpyxl        | 2.4.0           |          |
 +-----------------+-----------------+----------+
 
+.. _whatsnew_0230.api_breaking.dict_insertion_order:
+
+Creating dataframes and series from dicts preserves dict insertion order for python 3.6+
```
Comment (jschendel) [inline doc/source/whatsnew/v0.23.0.txt]: Some capitalization:  dataframes --> DataFrames, series --> Series, python --> Python

File: pandas/tests/series/test_constructors.py [hunk 12]
```diff
@@ -783,6 +783,18 @@ def test_constructor_dict(self):
         expected.iloc[1] = 1
         assert_series_equal(result, expected)
 
+    def test_constructor_dict_order(self):
```
Comment (jschendel) [inline pandas/tests/series/test_constructors.py]: Maybe could decorate these tests with `@pytest.mark.skipif(not PY36, reason="Python 3.6 specific behavior")`.  Not sure if we're gaining much testing the non-PY36 path.  Would wait to see what others think before making this change though.

File: doc/source/whatsnew/v0.23.0.txt [hunk 13]
```diff
@@ -320,6 +320,57 @@ If installed, we now require:
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Until Python 3.6, dicts in Python had no formally defined ordering. Python
+version 3.6 and later have changed the ordering definition of dicts, so dicts
+in these newer versions are ordered by insertion order
+(see also `PEP 468 <https://www.python.org/dev/peps/pep-0468/>`_).
+Pandas will from version 0.23 use insertion order, when creating series or
```
Comment (jschendel) [inline doc/source/whatsnew/v0.23.0.txt]: series --> Series

File: doc/source/whatsnew/v0.23.0.txt [hunk 14]
```diff
@@ -320,6 +320,57 @@ If installed, we now require:
...
+data frames from dicts (:issue:`19018`) .
```
Comment (jschendel) [inline doc/source/whatsnew/v0.23.0.txt]: data frames --> DataFrames

File: pandas/tests/series/test_constructors.py (hunk 12 above)
Comment (topper-123) [reply pandas/tests/series/test_constructors.py]: Pandas has clearly defined ordering for python < 3.6 also (sort  by values). So it's nneeded to test this distinction and not use a skipif, IMO.

**Comment by topper-123** [comment]
I've uploaded changes, incl. documentation changes.

File: doc/source/dsintro.rst [hunk 15]
```diff
@@ -81,9 +81,21 @@ index is passed, one will be created having values ``[0, ..., len(data) - 1]``.
 
 **From dict**
 
-If ``data`` is a dict, if **index** is passed the values in data corresponding
-to the labels in the index will be pulled out. Otherwise, an index will be
-constructed from the sorted keys of the dict, if possible.
+When creating a pandas Series from a dict, the Series will be ordered by the
```
Comment (jreback) [inline doc/source/dsintro.rst]: add a versionadded tag

File: doc/source/dsintro.rst [hunk 16]
```diff
@@ -277,6 +289,8 @@ The row and column labels can be accessed respectively by accessing the
 From dict of ndarrays / lists
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+The columns will be ordered by the dict insertion order, unless you're using
```
Comment (jreback) [inline doc/source/dsintro.rst]: add this in a ::note..

File: doc/source/dsintro.rst [hunk 17]
```diff
@@ -277,6 +289,8 @@ The row and column labels can be accessed respectively by accessing the
...
+Python version < 3.6, then the columns will be ordered lexically/alphabetically.
```
Comment (jreback) [inline doc/source/dsintro.rst]: just specify lexically

File: doc/source/whatsnew/v0.23.0.txt [hunk 18]
```diff
@@ -240,7 +240,7 @@ The :func:`DataFrame.assign` now accepts dependent keyword arguments for python
   using ``.assign()`` to update an existing column. Previously, callables
   referring to other variables being updated would get the "old" values
 
-  Previous Behaviour:
+  Previous behavior:
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: hmm, is the capitalization consistent across all of these entries? rather you leave this alone for now (and we can come back and clean it up later)

File: doc/source/whatsnew/v0.23.0.txt [hunk 19]
```diff
@@ -320,6 +320,57 @@ If installed, we now require:
 | openpyxl        | 2.4.0           |          |
 +-----------------+-----------------+----------+
 
+.. _whatsnew_0230.api_breaking.dict_insertion_order:
+
+Creating Dataframes and Series from dicts preserves dict insertion order for python 3.6+
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: make this shorter title

File: doc/source/whatsnew/v0.23.0.txt [hunk 20]
```diff
@@ -320,6 +320,57 @@ If installed, we now require:
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Until Python 3.6, dicts in Python had no formally defined ordering. Python
+version 3.6 and later have changed the ordering definition of dicts, so dicts
+in these newer versions are ordered by insertion order
+(see also `PEP 468 <https://www.python.org/dev/peps/pep-0468/>`_).
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: don't use parens in the 'see also', rather highlite some relevant phrase with this reference.

File: doc/source/whatsnew/v0.23.0.txt [hunk 21]
```diff
@@ -320,6 +320,57 @@ If installed, we now require:
...
+Pandas will from version 0.23 use insertion order, when creating Series or
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: you don't need from version 0.23, this is the whatsnew

File: doc/source/whatsnew/v0.23.0.txt [hunk 22]
```diff
@@ -320,6 +320,57 @@ If installed, we now require:
...
+DataFrames from dicts (:issue:`19018`) .
+
+Previous behavior (and current behavior if on Python < 3.6):
+
+.. code-block:: ipython
+
+   In [1]: pd.Series({'Income': 2000,
+   ...                 'Expenses': -1500,
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: this is not aligned

File: pandas/tests/series/test_constructors.py (hunk 12 above)
Comment (jreback) [reply pandas/tests/series/test_constructors.py]: @topper-123 let's go ahead and make 2 different tests here, 1 for >= 3.6 and one for less, they should skip on the alternative, same for the tests in frame.

File: pandas/core/sparse/frame.py [hunk 23]
```diff
@@ -138,7 +142,10 @@ def _init_dict(self, data, index, columns, dtype=None):
             columns = _ensure_index(columns)
             data = {k: v for k, v in compat.iteritems(data) if k in columns}
         else:
-            columns = Index(com._try_sort(list(data.keys())))
+            keys = list(data.keys())
```
Comment (jreback) [inline pandas/core/sparse/frame.py]: try pushing this entirely to ``_try_sort``, see if anything breaks as this used in a couple of places and will provide some 3.6 consistency.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/sparse/frame.py (hunk 23 above)
Comment (topper-123) [reply pandas/core/sparse/frame.py]: ``_try_sort`` is used for sorting lists in a couple of places.

I don't think it's a good idea to have a sort function that doesn't sort on ordered dict keys, but sorts on ordered lists. 

Maybe a differently named function, e.g. ``_dict_keys_to_ordered_list``?

Comment (jreback) [reply pandas/core/sparse/frame.py]: ok with renaming this, but do try to change all of the existing functions to use it.
You are creating an inconsistency by making dict inputs follow insertion order, rather have *all* instances fixed.

Comment (topper-123) [reply pandas/core/sparse/frame.py]: I've changed this for all places where ``_try_sort`` is used to sort  list of dict keys.

File: doc/source/dsintro.rst [hunk 24]
```diff
@@ -81,9 +81,26 @@ index is passed, one will be created having values ``[0, ..., len(data) - 1]``.
...
+
+   If you're using Python < 3.6 or Pandas < 0.23, and index is not passed,
+   the Series index will be the lexically ordered list of dict keys.
+
+.. ipython:: python
+
+   d = {'b' : 1, 'a' : 0, 'c' : 2}
+   pd.Series(d)
+
+If you in the example above were on a Python version lower than 3.6 or Pandas
```
Comment (TomAugspurger) [inline doc/source/dsintro.rst]: typo: I think "you in" can be removed

File: doc/source/whatsnew/v0.23.0.txt [hunk 25]
```diff
@@ -320,6 +320,57 @@ If installed, we now require:
...
 +-----------------+-----------------+----------+
 
+.. _whatsnew_0230.api_breaking.dict_insertion_order:
+
+Instantation from dicts preserves dict insertion order for python 3.6+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Until Python 3.6, dicts in Python had no formally defined ordering. Python
+version 3.6 and later have changed the ordering definition of dicts, so dicts
+in these newer versions are ordered by insertion order, see
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.23.0.txt]: Change the second sentence to "For Python 3.6 and later, dictionaries are ordered by insertion order."?

**Comment by topper-123** [comment]
The errors in travis-ci were

1:
One which passed locally, so sure ow to handle that. It complains about resources, so maybe the failures are related to that?

2:
One with the exit quote:
> The job exceeded the maximum time limit for jobs, and has been terminated.

so unrelated to my PR.

Maybe the tests should be run once again?

**Comment by topper-123** [comment]
The travis ci bug passes locally, but fails in travis ci. 

Anyone got an idea why this is failing?

**Comment by topper-123** [comment]
The test fails on

**Comment by topper-123** [comment]
Green.

**Review by TomAugspurger — APPROVED** [review]

File: doc/source/dsintro.rst [hunk 26]
```diff
@@ -81,9 +81,26 @@ index is passed, one will be created having values ``[0, ..., len(data) - 1]``.
 
 **From dict**
 
-If ``data`` is a dict, if **index** is passed the values in data corresponding
-to the labels in the index will be pulled out. Otherwise, an index will be
-constructed from the sorted keys of the dict, if possible.
+.. note::
```
Comment (jorisvandenbossche) [inline doc/source/dsintro.rst]: Can you move this note a bit below? It seems strange to start this "from dict" section with a note without first explaining what it actually is. 
So I would give an example first, and only afterwards noting that the output depends on the python version

File: doc/source/dsintro.rst [hunk 27]
```diff
@@ -243,12 +260,21 @@ not matching up to the passed index.
 If axis labels are not passed, they will be constructed from the input data
 based on common sense rules.
 
+.. note::
+
+   When the data is a dict, and columns is not passed, the DataFrame columns
```
Comment (jorisvandenbossche) [inline doc/source/dsintro.rst]: "and columns is not passed" -> I would try to make it clear that it is about a keyword `columns`, by either qouting it, or "no columns are passed", or "`columns` is not specified"

**Review by jorisvandenbossche — COMMENTED** [review]
small doc comment

**Comment by topper-123** [comment]
Comments from @jorisvandenbossche addressed.

**Comment by jorisvandenbossche** [comment]
@topper-123 For future reference, could you add commits instead of amending it to the previous or squashing? That makes it easier to see what you updated.

**Review by jorisvandenbossche — APPROVED** [review]
Changes look good!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @topper-123

**Pull request merged**